### PR TITLE
Additional code coverage in System.Text.Json - no public constructors

### PR DESCRIPTION
**Pull request by alanisaac**
Part of #32341

Added tests covering cases where deserializing collections without a public, parameterless constructor should throw a `NotSupportedException`.

### CODE REVIEW

File: src/libraries/System.Text.Json/tests/Serialization/Value.ReadTests.ConcurrentCollections.cs [hunk 1]
```diff
@@ -38,5 +38,12 @@ public static void Read_ConcurrentCollection_Throws()
             // Not supported. Not IList, and we don't detect the add method for this collection.
             Assert.Throws<NotSupportedException>(() => JsonSerializer.Deserialize<ConcurrentBag<string>>(@"[""1""]"));
         }
+
+        [Fact]
+        public static void Read_ConcurrentCollection_NoPublicConstructor_Throws()
+        {
+            Assert.Throws<NotSupportedException>(() => JsonSerializer.Deserialize<GenericConcurrentQueuePrivateConstructor<string>>(@"[""1""]"));
+            Assert.Throws<NotSupportedException>(() => JsonSerializer.Deserialize<GenericConcurrentStackPrivateConstructor<string>>(@"[""1""]"));
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/tests/Serialization/Value.ReadTests.ConcurrentCollections.cs]:
```suggestion
            ex = Assert.Throws<NotSupportedException>(() => JsonSerializer.Deserialize<GenericConcurrentStackPrivateConstructor<string>>(@"[""1""]"));
            Assert.Contains(typeof(GenericConcurrentStackPrivateConstructor).ToString(), ex.Message);
```

File: src/libraries/System.Text.Json/tests/Serialization/Value.ReadTests.ConcurrentCollections.cs [hunk 2]
```diff
@@ -38,5 +38,12 @@ public static void Read_ConcurrentCollection_Throws()
             // Not supported. Not IList, and we don't detect the add method for this collection.
             Assert.Throws<NotSupportedException>(() => JsonSerializer.Deserialize<ConcurrentBag<string>>(@"[""1""]"));
         }
+
+        [Fact]
+        public static void Read_ConcurrentCollection_NoPublicConstructor_Throws()
+        {
+            Assert.Throws<NotSupportedException>(() => JsonSerializer.Deserialize<GenericConcurrentQueuePrivateConstructor<string>>(@"[""1""]"));
```
[the remark this answers is not in the document]
Comment (alanisaac) [reply src/libraries/System.Text.Json/tests/Serialization/Value.ReadTests.ConcurrentCollections.cs]: fixed in 0397aa0

I also added the same assertions to other similar tests in 9d91217.

File: src/libraries/System.Text.Json/tests/Serialization/Value.ReadTests.GenericCollections.cs [hunk 3]
```diff
@@ -1161,5 +1179,41 @@ public static void Read_HigherOrderCollectionInheritance_Works()
             Assert.Equal("test", JsonSerializer.Deserialize<MyMyList<string>>(json).First());
             Assert.Equal("test", JsonSerializer.Deserialize<MyListString>(json).First());
         }
+
+        [Fact]
+        public static void Read_Generic_NoPublicConstructor_Throws()
```
Comment (layomia) [inline src/libraries/System.Text.Json/tests/Serialization/Value.ReadTests.GenericCollections.cs]: We should be able to use `[Theory]` + the non-generic `JsonSerializer.Deserialize` method here and elsewhere.

Comment (alanisaac) [reply src/libraries/System.Text.Json/tests/Serialization/Value.ReadTests.GenericCollections.cs]: Fixed in bccf92d

In a few tests, we reference data from fields of the form `public static readonly string s_json`, which cannot be used directly by `[InlineData]`, so I refactored to `[MemberData]` tests in 
921fc3a.  If we don't like the way `[MemberData]` works out for those tests, I can roll that back.

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests.csproj [hunk 4]
```diff
@@ -82,6 +82,7 @@
     <Compile Include="Serialization\SpanTests.cs" />
     <Compile Include="Serialization\Stream.ReadTests.cs" />
     <Compile Include="Serialization\Stream.WriteTests.cs" />
+    <Compile Include="Serialization\TestClasses.ConcurrentCollections.cs" />
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/tests/System.Text.Json.Tests.csproj]: nit: move below `TestClasses.cs`

File: src/libraries/System.Text.Json/tests/Serialization/TestClasses.NonGenericCollections.cs [hunk 5]
```diff
@@ -180,6 +185,11 @@ public IEnumerator GetEnumerator()
         }
     }
 
+    public class WrapperForICollectionPrivateConstructor : WrapperForICollection
+    {
+        private WrapperForICollectionPrivateConstructor() { }
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/tests/Serialization/TestClasses.NonGenericCollections.cs]: How about adding internal ctors as well?

File: src/libraries/System.Text.Json/tests/Serialization/Value.ReadTests.GenericCollections.cs (hunk 3 above)
Comment (layomia) [reply src/libraries/System.Text.Json/tests/Serialization/Value.ReadTests.GenericCollections.cs]: This change is fine.

**Comment by layomia** [comment]
Just a couple more comments @alanisaac, then this PR will be ready to merge.

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests.csproj (hunk 4 above)
Comment (alanisaac) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests.csproj]: fixed in 03ec773

File: src/libraries/System.Text.Json/tests/Serialization/TestClasses.NonGenericCollections.cs (hunk 5 above)
Comment (alanisaac) [reply src/libraries/System.Text.Json/tests/Serialization/TestClasses.NonGenericCollections.cs]: added classes and tests for internal ctors in fe06beb

**Review by layomia — APPROVED** [review]
Thanks!

**Pull request merged**